### PR TITLE
[Gene] Get the GeneVC hooked up with all native modules for Eigen

### PR DESCRIPTION
**Pull request by orta**
Will comment inline

### CODE REVIEW

File: Example/Emission/AppDelegate.m [hunk 1]
```diff
@@ -72,6 +73,7 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
   } else {
     [auth presentAuthenticationPromptOnViewController:rootVC completion:^{
       NSLog(@"Logged in successfully :)");
+      [self setupEmissionWithUserID:[auth userID] accessToken:[auth token] keychainService:service];
     }];
```
Comment (orta) [inline Example/Emission/AppDelegate.m]: missed this earlier, I even made the callback for it 👯‍♂️

File: Example/Emission/AppDelegate.m [hunk 2]
```diff
@@ -129,6 +131,27 @@ - (void)setupEmissionWithUserID:(NSString *)userID accessToken:(NSString *)acces
...
+    NSLog(@"Gene(%@).follow = %@", geneID, @(following));
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+      //      if (randomBOOL()) {
+      block(@[[NSNull null], @(following)]);
+      //      } else {
+      //        NSLog(@"Simulated follow request ‘failed’.");
+      //        block(@[RCTJSErrorFromNSError([NSError errorWithDomain:@"Artsy" code:42 userInfo:nil]), @(!following)]);
+      //      }
+    });
+  };
```
Comment (orta) [inline Example/Emission/AppDelegate.m]: Copy pasta of the Artist follow stuff

File: Example/Emission/AppDelegate.m [hunk 3]
```diff
@@ -156,8 +179,12 @@ - (void)setupEmissionWithUserID:(NSString *)userID accessToken:(NSString *)acces
   emission.eventsModule.eventOccurred = ^(UIViewController * _Nonnull fromViewController, NSDictionary * _Nonnull info) {
     NSLog(@"[Event] %@ - %@", fromViewController.class, info);
   };
-}
 
+  emission.refineModule.triggerRefine = ^(NSDictionary *_Nonnull metadata, UIViewController *_Nonnull controller, RCTPromiseResolveBlock resolve, RCTPromiseRejectBlock reject) {
+    sleep(1);
+    resolve(@{});
+  };
+}
```
Comment (orta) [inline Example/Emission/AppDelegate.m]: simplest for now

File: Example/Emission/AuthenticationManager.m [hunk 4]
```diff
@@ -16,9 +16,10 @@ @implementation AuthenticationManager
...
 - (instancetype)initWithService:(NSString *)service
 {
-  self = [super init];
-  _service = service;
-  [self updateFromStoredCredentials];
+  if ((self = [super init])) {
+    _service = service;
+    [self updateFromStoredCredentials];
+  };
   return self;
```
Comment (orta) [inline Example/Emission/AuthenticationManager.m]: feedback from last PR

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="true">No CHANGELOG added.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">The following files do not have tests: <a href="https://github.com/artsy/emission/blob/767e9352ecde865353a97a69bdf379930e909af7/lib/components/artist/header.js">lib/components/artist/header.js</a>, <a href="https://github.com/artsy/emission/blob/767e9352ecde865353a97a69bdf379930e909af7/lib/components/gene/header.js">lib/components/gene/header.js</a>, <a href="https://github.com/artsy/emission/blob/767e9352ecde865353a97a69bdf379930e909af7/lib/containers/gene.js">lib/containers/gene.js</a> &amp; <a href="https://github.com/artsy/emission/blob/767e9352ecde865353a97a69bdf379930e909af7/lib/native_modules/refine_callback.js">lib/native_modules/refine_callback.js</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: Pod/Classes/EigenCommunications/ARRefineOptionsModule.m [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+#import <React/UIView+React.h>
+#import <React/RCTRootView.h>
+
+@implementation ARRefineOptionsModule
+
+@synthesize bridge = _bridge;
+
+RCT_EXPORT_MODULE();
+
+RCT_EXPORT_METHOD(triggerRefinePanel:(nonnull NSNumber *)reactTag metadata:(nonnull NSDictionary *)info resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
```
Comment (orta) [inline Pod/Classes/EigenCommunications/ARRefineOptionsModule.m]: exposes a promises

File: Pod/Classes/TemporaryAPI/ARTemporaryAPIModule.m [hunk 6]
```diff
@@ -9,9 +9,19 @@ @implementation ARTemporaryAPIModule
   self.artistFollowStatusProvider(artistID, block);
 }
 
-RCT_EXPORT_METHOD(setFollowStatus:(BOOL)status artistID:(NSString *)artistID completion:(RCTResponseSenderBlock)block)
+RCT_EXPORT_METHOD(setFollowArtistStatus:(BOOL)status artistID:(NSString *)artistID completion:(RCTResponseSenderBlock)block)
```
Comment (orta) [inline Pod/Classes/TemporaryAPI/ARTemporaryAPIModule.m]: _note this_ - the API inside JS would be `setFollowStatus` but you can't have two of them, so I gave them unique names

File: lib/components/artist/header.js [hunk 7]
```diff
@@ -45,7 +45,7 @@ class Header extends React.Component {
 
   handleFollowChange = () => {
     const newFollowersCount: number = this.state.following ? (this.state.followersCount - 1) : (this.state.followersCount + 1)
-    ARTemporaryAPIModule.setFollowStatus(!this.state.following, this.props.artist._id, (error, following) => {
+    ARTemporaryAPIModule.setFollowArtistStatus(!this.state.following, this.props.artist._id, (error, following) => {
```
Comment (orta) [inline lib/components/artist/header.js]: update for the above change

File: lib/components/gene/header.js [hunk 8]
```diff
@@ -50,14 +77,11 @@ class Header extends React.Component {
...
         <View style={styles.followButton}>
-          <InvertedButton text="Follow" textOpacity={0.5} />
+          <InvertedButton text="" onPress={this.handleFollowChange} />
         </View>
       )
     }
   }
-
-// NO-OP for now
-  handleFollowChange = () => {}
```
Comment (orta) [inline lib/components/gene/header.js]: this caught me for ages, a silent overwrite of my new function

File: lib/native_modules/refine_callback.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  let reactTag
+  try {
+    reactTag = findNodeHandle(component)
+  } catch (err) {
+    console.error("could not find tag")
+    return
+  }
+  console.log("triggerRefine")
+  console.log(ARRefineOptionsModule)
+  return ARRefineOptionsModule.triggerRefinePanel(reactTag, info)
```
Comment (orta) [inline lib/native_modules/refine_callback.js]: logssss

**Comment by orta** [comment]
@mennenia just assigned you ^

**Comment by mennenia** [comment]
on it

File: Example/Emission.xcodeproj/xcshareddata/xcschemes/Emission.xcscheme [hunk 10]
```diff
@@ -88,6 +88,11 @@
             isEnabled = "YES">
          </EnvironmentVariable>
          <EnvironmentVariable
+            key = "SRCROOT"
+            value = "$(SRCROOT)"
+            isEnabled = "YES">
+         </EnvironmentVariable>
+         <EnvironmentVariable
```
Comment (orta) [inline Example/Emission.xcodeproj/xcshareddata/xcschemes/Emission.xcscheme]: https://github.com/artsy/artsy.github.io/pull/285

File: Example/Emission/ARRootViewController.m [hunk 11]
```diff
@@ -84,8 +84,7 @@ - (ARCellData *)jumpToArtist
 
 - (ARCellData *)jumpToRandomArtist
 {
-  NSDictionary *metadata = [[NSBundle mainBundle] infoDictionary];
-  NSString *sourceRoot = [metadata objectForKey:@"SourceRoot"];
+  NSString *sourceRoot = [NSProcessInfo processInfo].environment[@"SRCROOT"];
```
Comment (mennenia) [inline Example/Emission/ARRootViewController.m]: was this for the simulator?

File: Example/Emission/AppDelegate.m [hunk 12]
```diff
@@ -129,6 +131,27 @@ - (void)setupEmissionWithUserID:(NSString *)userID accessToken:(NSString *)acces
...
+    NSLog(@"Gene(%@).follow => %@", geneID, following);
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+      block(@[[NSNull null], following]);
+    });
+  };
+
+  emission.APIModule.geneFollowStatusAssigner = ^(NSString *geneID, BOOL following, RCTResponseSenderBlock block) {
+    NSLog(@"Gene(%@).follow = %@", geneID, @(following));
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+      //      if (randomBOOL()) {
```
Comment (mennenia) [inline Example/Emission/AppDelegate.m]: @alloy have we decided to completely get rid of the random bool?

File: Pod/Classes/TemporaryAPI/ARTemporaryAPIModule.m (hunk 6 above)
Comment (mennenia) [inline Pod/Classes/TemporaryAPI/ARTemporaryAPIModule.m]: 👍

File: Example/Emission/ARRootViewController.m (hunk 11 above)
Comment (mennenia) [inline Example/Emission/ARRootViewController.m]: nvm this is probably explained in the post (which I opened in a new tab, yet to read :P)

File: Example/Emission/AppDelegate.m (hunk 12 above)
Comment (alloy) [inline Example/Emission/AppDelegate.m]: Yeah it’s probably not worth it to make this usable 👍

File: Example/Emission/AppDelegate.m (hunk 3 above)
Comment (alloy) [inline Example/Emission/AppDelegate.m]: Out of curiosity, was there any reason to not do a delay with GCD?

File: Example/Emission/AuthenticationManager.m (hunk 4 above)
Comment (alloy) [inline Example/Emission/AuthenticationManager.m]: 👍 🙏

**Pull request merged**